keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -514,310 +481,126 @@ spec:
-    - name: driverVersion
-      string: 1.2.3
-    - name: runtimeVersion
-      string: 11.1.42
-    - name: memory
-      quantity: 16Gi
+    - name: manufacturer # a vendor-specific attribute, automatically qualified as cards.dra.example.com/manufacturer
+      string: Cards-are-us Inc.
     - name: productName:
-      string: ACME T1000 32GB
-  - name: gpu-1
-    attributes:
-    - name: UUID
-      string: GPU-6aa0af9e-a2be-88c8-d2b3-2240d25318d7
+      string: ACME T1000 16GB
     - name: driverVersion
-      string: 1.2.3
+      version: 1.2.3
     - name: runtimeVersion
       string: 11.1.42
     - name: memory
-      quantity: 32Gi
-    - name: productName:
-      string: ACME A4-PCIE-40GB
+      quantity: 16Gi
+    - name: powerSavingSupported
+      bool: true
+    - name: dra.k8s.io/pciRoot # a fictional standardized attribute, not actually part of this KEP
+      string: pci-root-0
 ```
 
-Where "gpu-0" represents one type of card and "gpu-1" represents another (with
-the attributes hanging off each serving to "define" their individual
-properties).
-
-Compared to labels, attributes in this model have values of exactly one type. As
+Compared to labels, attributes have values of exactly one type. As
 described later on, these attributes can be used in CEL expressions to select a
 specific resource for allocation on a node.
 
-While this model is still hypothetical, we do imagine real-world models
-attaching attributes to their resources in a similar way. To avoid any future
-conflicts, we plan to reserve any attributes with the ".k8s.io" suffix for
-future use and standardization by Kubernetes. This could be used to describe
+To avoid any future conflicts, we reserve any attributes with the ".k8s.io" domain suffix
+for future use and standardization by Kubernetes. This could be used to describe
 topology across resources from different vendors, for example, but this is out-
 of-scope for now.
 
-**Note:** If a driver needs to reduce resource capacity, then there is a risk
-that a claim gets allocated using that capacity while the kubelet is updating a
-`ResourceSlice`. The implementations of structured models must handle
-scenarios where more resources are allocated than available. The kubelet plugin
-of a DRA driver ***must*** double-check that the allocated resources are still
+**Note:** If a driver needs to remove a device or change its attributes,
+then there is a risk that a claim gets allocated based on the old
+`ResourceSlice`. The scheduler must handle
+scenarios where more devices are allocated than available. The kubelet plugin
+of a DRA driver ***must*** double-check that the allocated devices are still
 available when NodePrepareResource is called. If not, the pod cannot start until
-the resource comes back. Treating this as a fatal error during pod admission
+the device comes back. Treating this as a fatal error during pod admission
 would allow us to delete the pod and trying again with a new one.
```

Comment:
At present we don't do anything for DRA during pod admission. Everything is done at container start. We wanted to do it at pod-admission, but there is no `retry` loop during pod-admission.  We need to be able to handle transient failures in calls to `NodePrepareResources` and `NodeUnprepareResources()`, so until there is a way to do this in the pod-admission loop, we are stuck doing things at container start.

Reply:
Do we want to document that here?

Reply:
I only mention it because the text says:

>  Treating this as a fatal error during pod admission would allow us to delete the pod and trying again with a new one.

And my point is that we never look at it during pod admission (at least not currently).

Reply:
So that should say "Checking it at admission time and ... would allow us...".